tasks: Add interactive-bots.py script.

derAnfaenger opened the pull request:
* I set myself as the mentor for the bot tasks for now. @eeshangarg @abhijeetkaur feel free to add your emails as well, if you want.

* I added some initial dieas for creating bots and a list of bots to improve. These should be reviewed.

* Updated task descriptions, should be reviewed.

* I left the limit for tasks A & B at 30 for now, but I don't see why we shouldn't increase them to 100.

timabbott commented on scripts/2017/interactive-bots.py:
```diff
@@ -0,0 +1,117 @@
+args = parser.parse_args()
+
+bots_to_create = [
+    ('Chess', ('lets you play chess against the computer or other people on Zulip')),
```
I guess for this, we should make really sure that the student uses an existing chess engine library ; otherwise it'll be way too much work.  I think we should somehow put that into the description.

timabbott commented on scripts/2017/interactive-bots.py:
```diff
@@ -0,0 +1,117 @@
+    ('Four in a row', ('lets you play four in a row against the computer or other people on Zulip')),
+    ('Google Hangouts', ('posts a link to a Hangouts call.')),
+    ('lunch', ('lets users create and join lunch events')),
+    ('URL shortener', ('takes a link and sends back a shortened version using a shortener service')),
```
In terms of project priorities, I think we care a lot more about implementing Zulip bot versions of popular hubot/slack/errbot integrations (especially hubot, since it's kinda a pain), than writing new games.  There's 2 reasons for this:
* We need to make fewer decisions when porting an existing bot implementation, so less mentor coordination is required.
* The results will be bots we're confident our users will find useful, since they replicate functionality available with other chat apps.

So to make the best use of everyone's time, I think we should probably make this list be 80% "integrations we expect people to use" and 20% "write a game", rather than the other way around.  

Can you browser around and try to make this list have another 15 things chosen because they're popular integrations elsewhere?

derAnfaenger replied:
Yeah, will add some more. I also have a bunch of integrations on my list left, but these are more fitting for the "webhook integrations" task.

timabbott replied:
Cool.  It'd be reasonable for you to do a pass on the "webhook integrations" script to fill those in now, even though we haven't finished updating the task yet.

timabbott commented:
BTW, what's the story with `external_url`?  I'm just wondering if we want to have that point to the task description rather than the repo?  If so, there's no point in adding a default that's wrong, is there?

Also, for updates to scripts/, I think we might want to commit the version from last year and then a commit doing the changes on top, to make it easier to read the diff of what changed...

derAnfaenger commented:
For `external_url` I thought it to be "righter" to refer to the repo which gives the student a README.md with a general overview, rather than referring to nothing. But we can make it required as well, to ensure that every updated script thinks about what to put there.

derAnfaenger commented:
Updated with:

* making `external_url` a required argument without default value
* making all arguments to `upload_task` keyword-only (will help for the new parameter and changing API in general)
* adding the translation task description in `external_url`
* one can now see the diff for `interactive-bots.py`
* researched several integrations for bots and incoming webhooks, and updated the respective lists. Some special cases:
  * Intercom is the company that provides the little chat icons you can now see everywhere. I added it to the "incoming webhook" task list because it provides webhooks. It's also on the "bots" task lists, since it'd be nice to have a bot that allows people to actually respond over Zulip, not just receive messages.
  * `statsbot` would make a good integration, but you have to request API access. I didn't think it'd make a good GCI task. Let me know what you think.
  * `Mailchimp exists in our "legacy hubot adapter" integrations. Not sure how to go about that.

timabbott commented:
I think we'd be happy to replace any "legacy hubot adapter" integrations with a direct one, since hubot isn't the best.  

Merged, thanks @derAnfaenger!  I think for `statsbot`, we could try asking for access and see how long it takes, but yeah, seems bad as an early-in-GCI task.
